File: adaptors/python/json_storage_connection.py
import socket
import json
from typing import Any, Optional, Union
from config.config_manager import ConfigManager

class JsonStorageConnection:
    def __init__(self, host: str = 'localhost', port: Optional[int] = None):
        self.host = host
        self.port = port or int(ConfigManager.get('PORT', 5555))
        self.socket: Optional[socket.socket] = None
        self.connected = False
        self.authenticated = False
        self.debug = True
        self.timeout = 5
        
    def connect(self) -> bool:
        if self.debug:
            print("Creating socket...")
            
        self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.socket.settimeout(self.timeout)
        
        try:
            self.socket.connect((self.host, self.port))
            self.connected = True
            self._handle_authentication()
            return True
        except Exception as e:
            raise ConnectionError(f"Failed to connect: {e}")
            
    def _handle_authentication(self) -> bool:
        if self.authenticated:
            return True
            
        response = self._read()
        if response.strip() != "AUTH_REQUIRED":
            self.authenticated = True
            return True
            
        password = ConfigManager.get('SERVER_PASSWORD', '').strip("'\"")
        if not password:
            raise Exception("Authentication required but SERVER_PASSWORD not configured")
            
        if self.debug:
            print("Attempting authentication...")
            
        self._write(f"AUTH {password}")
        response = self._read()
        
        if "OK" not in response:
            raise Exception(f"Authentication failed: {response.strip()}")
            
        self.authenticated = True
        return True
        
    def _write(self, data: str) -> None:
        if not self.socket:
            raise Exception("Not connected")
        self.socket.send(f"{data}\n".encode())
        
    def _read(self) -> str:
        if not self.socket:
            raise Exception("Not connected")
        return self.socket.recv(1024).decode()
        
    def set(self, key: str, value: Any, ttl: int = -1) -> bool:
        if not self.connected:
            raise Exception("Not connected")
            
        if isinstance(value, (dict, list)):
            value = json.dumps(value)
        
        command = f"SET {key} {value}"
        self._write(command)
        response = self._read()
        
        if response.strip() != 'OK':
            return False
            
        if ttl > 0:
            self._write(f"EXPIRE {key} {ttl}")
            expire_response = self._read()
            return expire_response.strip() == 'OK'
            
        return True
        
    def get(self, key: str) -> Any:
        if not self.connected:
            raise Exception("Not connected")
            
        if key == "PING":
            self._write("PING")
            return self._read().strip()
            
        self._write(f"GET {key}")
        response = self._read().strip()
        
        if response == 'nil':
            return None
            
        response = response.strip('"')
        
        if response == 'null':
            return None
        elif response == 'true':
            return True
        elif response == 'false':
            return False
            
        try:
            if response.startswith('{') or response.startswith('['):
                return json.loads(response)
        except:
            pass
            
        try:
            if '.' in response:
                return float(response)
            return int(response)
        except:
            pass
            
        return response
        
    def ttl(self, key: str) -> int:
        if not self.connected:
            raise Exception("Not connected")
            
        self._write(f"TTL {key}")
        response = self._read().strip()
        
        if response == 'nil':
            return -2
            
        return int(response)
        
    def delete(self, key: str) -> bool:
        if not self.connected:
            raise Exception("Not connected")
            
        self._write(f"DEL {key}")
        response = self._read().strip()
        return response == "1"
        
    def close(self) -> None:
        if self.socket and self.connected:
            self.socket.close()
            self.connected = False
            self.socket = None

File: adaptors/nodejs/JsonStorageConnection.js
const net = require('net');
const { ConfigManager } = require('./config/ConfigManager');

class JsonStorageConnection {
    constructor(host = 'localhost', port = null) {
        this.host = host;
        this.port = port ?? ConfigManager.get('PORT', 5555);
        this.socket = null;
        this.connected = false;
        this.authenticated = false;
        this.debug = true;
        this.timeout = 5000; // milliseconds
        this.maxListeners = 20;
    }

    async connect() {
        if (this.debug) {
            console.log(`Creating socket connection to ${this.host}:${this.port}...`);
        }

        return new Promise((resolve, reject) => {
            this.socket = new net.Socket();
            
            this.socket.setTimeout(this.timeout);

            this.socket.setMaxListeners(this.maxListeners);

            this.socket.on('timeout', () => {
                reject(new Error('Socket timeout'));
                this.close();
            });

            this.socket.connect(this.port, this.host, async () => {
                this.connected = true;
                try {
                    await this.handleAuthentication();
                    resolve(true);
                } catch (error) {
                    this.close();
                    reject(error);
                }
            });

            this.socket.on('error', (error) => {
                reject(error);
                this.close();
            });
        });
    }

    async handleAuthentication() {
        if (this.authenticated) return true;

        const response = await this.read();
        if (response.trim() !== 'AUTH_REQUIRED') {
            this.authenticated = true;
            return true;
        }

        const password = ConfigManager.get('SERVER_PASSWORD')?.replace(/['"]/g, '');
        if (!password) {
            throw new Error('Authentication required but SERVER_PASSWORD not configured');
        }

        if (this.debug) {
            console.log('Attempting authentication...');
        }

        await this.write(`AUTH ${password}`);
        const authResponse = await this.read();

        if (!authResponse.includes('OK')) {
            throw new Error(`Authentication failed: ${authResponse.trim()}`);
        }

        this.authenticated = true;
        return true;
    }

    verifyAuthentication() {
        if (!this.authenticated) {
            throw new Error('Not authenticated with server');
        }
    }

    write(data) {
        return new Promise((resolve, reject) => {
            if (!this.socket) {
                reject(new Error('Socket not connected'));
                return;
            }

            this.socket.write(data + '\n', (error) => {
                if (error) reject(error);
                else resolve();
            });
        });
    }

    read() {
        return new Promise((resolve, reject) => {
            if (!this.socket) {
                reject(new Error('Socket not connected'));
                return;
            }

            const dataHandler = (data) => {
                cleanup();
                resolve(data.toString().trim());
            };

            const errorHandler = (error) => {
                cleanup();
                reject(error);
            };

            const cleanup = () => {
                this.socket.removeListener('data', dataHandler);
                this.socket.removeListener('error', errorHandler);
            };

            this.socket.once('data', dataHandler);
            this.socket.once('error', errorHandler);
        });
    }

    async set(key, value, ttl = -1) {
        if (!this.connected) throw new Error('Not connected');
        this.verifyAuthentication();

        // Handle different data types
        let processedValue;
        if (typeof value === 'string') {
            // For strings, escape quotes and wrap in quotes
            processedValue = `"${value.replace(/"/g, '\\"')}"`;
        } else if (typeof value === 'boolean') {
            processedValue = value ? 'true' : 'false';
        } else if (value === null) {
            processedValue = 'null';
        } else if (typeof value === 'object') {
            processedValue = JSON.stringify(value);
        } else if (typeof value === 'number') {
            processedValue = value.toString();
        }

        const command = `SET ${key} ${processedValue}`;
        
        if (this.debug) {
            console.log(`Sending command: ${command}`);
            console.log('Processed value:', processedValue);
        }

        await this.write(command);
        const response = await this.read();

        if (response.trim() !== 'OK') {
            return false;
        }

        if (ttl > 0) {
            await this.write(`EXPIRE ${key} ${ttl}`);
            const expireResponse = await this.read();
            return expireResponse.trim() === 'OK';
        }

        return true;
    }

    async get(key) {
        if (!this.connected) throw new Error('Not connected');
        this.verifyAuthentication();

        if (key === 'PING') {
            await this.write('PING');
            return await this.read();
        }

        await this.write(`GET ${key}`);
        const response = await this.read();

        if (response === 'nil') return null;

        // Remove surrounding quotes if present
        if (response.startsWith('"') && response.endsWith('"')) {
            return response.slice(1, -1).replace(/\\"/g, '"');
        }

        // Handle other types
        if (response === 'true') return true;
        if (response === 'false') return false;
        if (response === 'null') return null;
        
        // Try parsing as number
        if (!isNaN(response)) {
            return response.includes('.') ? parseFloat(response) : parseInt(response);
        }
        
        // Try parsing as JSON for objects and arrays
        try {
            if ((response.startsWith('{') && response.endsWith('}')) ||
                (response.startsWith('[') && response.endsWith(']'))) {
                return JSON.parse(response);
            }
        } catch (e) {
            // If JSON parsing fails, return as is
        }

        return response;
    }

    async ttl(key) {
        if (!this.connected) throw new Error('Not connected');
        this.verifyAuthentication();

        await this.write(`TTL ${key}`);
        const response = await this.read();

        if (response === 'nil') return -2;
        return parseInt(response);
    }

    async del(key) {
        if (!this.connected) throw new Error('Not connected');
        this.verifyAuthentication();

        await this.write(`DEL ${key}`);
        const response = await this.read();

        return response === '1';
    }

    close() {
        if (this.socket && this.connected) {
            this.socket.destroy();
            this.connected = false;
            this.socket = null;
        }
    }
}

module.exports = { JsonStorageConnection };